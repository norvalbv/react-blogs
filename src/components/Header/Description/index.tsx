import React, { ReactElement } from 'react';

export type DescriptionProps = {
  children?: string | JSX.Element;
  className?: string;
  testId?: string;
};

const Description = ({ children, className, testId }: DescriptionProps): ReactElement | null => {
  if (!children) {
    return null;
  }

  if (typeof children === 'string') {
    return (
      <p className={className} data-testid={testId || `${children} Description`}>
        {children}
      </p>
    );
  }

  return children;
};

export default Description;
