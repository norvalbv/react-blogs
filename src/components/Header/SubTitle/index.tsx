import React, { ReactElement } from 'react';
import { styles, themes } from 'styles/themes.css';
import { getClassName } from 'utils';

export type SubtitleProps = {
  children?: string | ReactElement;
  className?: string;
  props?: unknown;
};

const Subtitle = ({ children, className, ...props }: SubtitleProps): ReactElement | null => {
  if (!children) {
    return null;
  }

  if (typeof children === 'string') {
    return (
      <h4
        className={
          className || getClassName('h4') || `${themes.SHADES_OF_GREEN.nodes} ${styles.h1}`
        }
        {...props}
      >
        {children}
      </h4>
    );
  }

  return children;
};

export default Subtitle;
