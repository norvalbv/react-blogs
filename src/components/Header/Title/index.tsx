import React, { ReactElement } from 'react';
import { DefTheme } from 'types';
import { getClassName, isLabelInProps } from 'utils';

export type TitleProps = {
  children?: string | ReactElement;
  className?: string;
  /**
   * Default 2.
   */
  level?: 1 | 2 | 3;
  theme?: DefTheme;
  props?: unknown;
};

const Title = ({
  children,
  className,
  level = 2,
  theme,
  ...props
}: TitleProps): ReactElement | null => {
  if (!children) {
    return null;
  }

  if (level === 1) {
    return (
      <h1 className={getClassName({ tag: 'h1', theme, className })} {...isLabelInProps(props)}>
        {children}
      </h1>
    );
  }

  if (level === 2) {
    return (
      <h2 className={getClassName({ tag: 'h2', theme, className })} {...isLabelInProps(props)}>
        {children}
      </h2>
    );
  }

  return (
    <h3 className={getClassName({ tag: 'h3', theme, className })} {...isLabelInProps(props)}>
      {children}
    </h3>
  );
};

export default Title;
