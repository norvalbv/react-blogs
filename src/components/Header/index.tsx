import Subtitle from 'components/Header/SubTitle';
import Title from 'components/Header/Title';
import React, { ReactElement } from 'react';
import Description, { DescriptionProps } from './Description';
import { SubtitleProps } from './SubTitle';
import { TitleProps } from './Title';

export type HeaderProps = {
  // Own props
  className?: string;
  title?: Omit<TitleProps, 'children'> & { text: string };
  subtitle?: Omit<SubtitleProps, 'children'> & { text: string | ReactElement };
  description?: Omit<DescriptionProps, 'children'> & { text: string | ReactElement };
};

const Header = ({ className, description, subtitle, title }: HeaderProps): ReactElement | null => {
  if (!title && !subtitle) {
    return null;
  }

  return (
    <header className={className}>
      <section className="flex w-full flex-col gap-1">
        <Title {...title}>{title?.text}</Title>
        <Subtitle {...subtitle}>{subtitle?.text}</Subtitle>
        <Description {...description}>{description?.text}</Description>
      </section>
    </header>
  );
};

export default Header;
