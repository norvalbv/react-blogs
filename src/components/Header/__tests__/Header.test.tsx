import React, { FC } from 'react';
import renderer from 'react-test-renderer';

import Header, { HeaderProps } from '..';

const HeaderWithAllProps: FC<Required<HeaderProps>> = Header;

describe('<Header />', () => {
  test('should render nothing if title and subtitle are not provided', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with title provided', () => {
    const tree = renderer.create(<Header title="Test title" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with subtitle provided', () => {
    const tree = renderer.create(<Header subtitle="Test subtitle" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with all props', () => {
    const tree = renderer
      .create(
        <HeaderWithAllProps
          className="grid grid-cols-1 gap-1"
          title="Test title"
          titleClassName="text-h2"
          subtitle="Test subtitle"
          subtitleClassName="text-sm"
          description="Test Description"
          descriptionClassName="text-neutral-N00"
          animation={false}
          level={2}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
