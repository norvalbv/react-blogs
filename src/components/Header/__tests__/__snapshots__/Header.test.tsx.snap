// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render correctly with all props 1`] = `
<header
  className="grid grid-cols-1 gap-1"
>
  <h2
    className="text-2xl"
    data-testid="Test title Subtitle"
  >
    Test title
  </h2>
  <h4
    className="text-sm"
    data-testid="Test subtitle Subtitle"
  >
    Test subtitle
  </h4>
  <p
    className="text-neutral-500"
    data-testid="Test Description Description"
  >
    Test Description
  </p>
</header>
`;

exports[`<Header /> should render correctly with subtitle provided 1`] = `
<header>
  <h4
    className="themes_styles_h4__xd3hr9o defaults_defaults_h4__cxs0rs3"
    data-testid="Test subtitle Subtitle"
  >
    Test subtitle
  </h4>
</header>
`;

exports[`<Header /> should render correctly with title provided 1`] = `
<header>
  <h2
    className="themes_styles_h2__xd3hr9m defaults_defaults_h2__cxs0rs1"
    data-testid="Test title Subtitle"
  >
    Test title
  </h2>
</header>
`;

exports[`<Header /> should render nothing if title and subtitle are not provided 1`] = `null`;
